guest-js/index.ts

```diff
@@ -97,101 +121,215 @@ export class DownloadEventManager {
 }
 
 /**
- * Represents a download item with methods to control its lifecycle.
- * This class wraps a download item and provides methods to start, cancel, pause, resume,
- * and listen for changes to the download.
+ * Base class for all download states.
+ * Contains common properties and the listen method.
  */
-export class Download implements DownloadItem {
-   public key: string;
-   public url: string;
-   public path: string;
-   public progress: number;
-   public state: DownloadState;
+abstract class DownloadBase implements DownloadItem {
+   public readonly key: string;
+   public readonly url: string;
+   public readonly path: string;
+   public readonly progress: number;
+   public abstract readonly state: DownloadState;
 
-   public constructor(item: DownloadItem) {
+   protected constructor(item: DownloadItem) {
       this.key = item.key;
       this.url = item.url;
       this.path = item.path;
       this.progress = item.progress;
-      this.state = item.state;
    }
 
    /**
-   * Starts the download.
+   * Listen for changes to the download.
+   * To avoid memory leaks, the `unlisten` function returned by the promise
+   * should be called when no longer required.
+   * @param onChanged - Callback function invoked when the download has changed.
+   * @returns A promise with a function to remove the download listener.
+   *
+   * @example
+   * ```ts
+   * const unlisten = await download.listen((updatedDownload) => {
+   *   console.log('Download:', updatedDownload);
+   *   if (updatedDownload.state === DownloadState.PAUSED) {
+   *     updatedDownload.resume(); // TypeScript knows this is valid
+   *   }
+   * });
+   *
+   * // To stop listening
+   * unlisten();
+   * ```
+   */
+   public async listen(listener: (download: Download) => void): Promise<UnlistenFn> {
```

Comment:
ADIRL: do not allow `listen` for downloads that will never have updates, such as as those with status `Cancelled` and `Completed`